### PR TITLE
Completed Scrabble project

### PR DESCRIPTION
**Pull request by sallyamoore**
I completed breakfast on this one and started dabbling in lunch. I didn't finish lunch. I ran into an issue with the total_score method in the Player class -- I wanted to use the score method and constants from the Scrabble class rather than recreating that functionality, but I didn't have time to figure out how best to do that. What would you suggest? 

(Please excuse the typos in the messages that accompany some of my commits -- I'm having an issue with word wrapping in my terminal that makes it difficult to see these kinds of typos. I'll fix it before the next project.)

### CODE REVIEW

File: lib/scrabble-project/player.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    def initialize(name)
+      @name = name
+      @plays = []
+      @won = false
+      @total_score = 0
+    end
+
+    def play(word)
+      @plays.push(word)
+      return @plays
```
Comment (Elffers) [inline lib/scrabble-project/player.rb]: Quick style note: Since Ruby returns the last line of a method by default, you don't need to explicitly say `return` here.

File: lib/scrabble-project/scrabble-class.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    LETTERS = ("a".."z").to_a
+
+    def self.get_points(word)
+      # set initial value of score
+      score = 0
+
+      # convert to string, make letters lower case, & push letters to array
+      word_array = word.to_s.downcase.split("").push
```
Comment (Elffers) [inline lib/scrabble-project/scrabble-class.rb]: Couple things:
1) `word` should already be a string, so there's no need to call `.to_s` on it. If you wanted to ensure that the argument passed into `.get_points` is a string, you might consider putting it in your `.valid_input?` method.

2) Since `.split("")` already returns an array, you don't need `.push` at the end, here, especially if you're not providing an argument to push. In other words,  you're already assigning `word_array` to the return value of `word.downcase.split("")`.  On the other hand, if you wanted to start off with `word_array` as an empty array, how would you use `.push` to populate that array?

3) Also, there is another method on String that does exactly what `.split("")` does, and does not require an argument. Can you find it in the documentation?

File: lib/scrabble-project/scrabble-class.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    def self.valid_input?(word)
+      word = word.to_s.downcase
+
+      if (word == "") || (word == nil) || (word.length > 7) || (word.split("").all? { |letter| !LETTERS.include?(letter) })
+        valid = false
+      else
+        valid = true
+      end
+    end
+
```
Comment (Elffers) [inline lib/scrabble-project/scrabble-class.rb:47]: 👏 on extracting out `.valid_input?` and `.get_score` into a separate methods!

File: lib/scrabble-project/scrabble-class.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      # otherwise, return the shortest word in group.
+      else
+        return max_group.min_by { |word| word.length }
+      end
+
+    end
+
+    def self.highest_score_from(array_of_words)
+      if array_of_words.all? { |word| self.valid_input?(word)}
+        # if self.valid_input?(word) == true
```
Comment (Elffers) [inline lib/scrabble-project/scrabble-class.rb:82]: Great use of `.all?` 👍

File: lib/scrabble-project/scrabble-class.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        self.get_points(word)
+      end
+    end
+
+    def self.highest(array_of_words)
+
+      # score words and group by score value.
+      grouped_hash = array_of_words.group_by do |word|
+          score(word) end
+
```
Comment (Elffers) [inline lib/scrabble-project/scrabble-class.rb:65]: Style nit: Usually, the `end` should go on its own line after the method. If your block is short, try using curly braces instead, e.g. `{ |word| score(word) }`

File: lib/scrabble-project/player.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    end
+
+    def total_score     # This method doesn't work! I got as far as making it, but not as far as making it work. 
+      @plays.each do |word|
+      @total_score += (Scrabble::Scrabble.score(word))
+      end
+      return @total_score
+    end
```
Comment (Elffers) [inline lib/scrabble-project/player.rb:23]: You're on the right track! Think about it this way: Since `@total_score` is an instance variable, its state gets stored on the instance of Player. In other words, if you call the method `total_score` on a Player instance, it will sum up all the points currently in `@plays` and store it in `@total_score`. But what happens if the player has a few more plays, and then you call `total_score` again? 

Hint: Do you need `@total_score` as an instance variable at all?

File: spec/lib/player_spec.rb [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+      expect(player).to respond_to :total_score
+    end
+
+    # # I didn't get total_score working. This spec doesn't pass, so I commented it out for now. 
+    # let(:plays) { Scrabble::Player.plays(["hello", "medic", "qat"]) }
+    #
+    # it "sums word scores to create 'total_score'" do
+    #   expect(player.total_score).to eq(27)
+    # end
+
```
Comment (Elffers) [inline spec/lib/player_spec.rb:43]: Tip: If you want to ignore a test for now, rather than commenting out the whole test you can say `xit` instead of `it` (here, on line 40) and Rspec will mark the test as pending.

File: spec/lib/player_spec.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    it "responds to 'play'" do
+      expect(player).to respond_to :play
+    end
+
+    it "responds to 'total_score'" do
+      expect(player).to respond_to :total_score
+    end
+
+    # # I didn't get total_score working. This spec doesn't pass, so I commented it out for now. 
+    # let(:plays) { Scrabble::Player.plays(["hello", "medic", "qat"]) }
```
Comment (Elffers) [inline spec/lib/player_spec.rb:38]: Be careful of the way you set up your test. `.plays` is an instance variable on `Scrabble::Player`, so you'll get an error if you call it directly on the class. Also, since it is an instance variable, it doesn't take any arguments.

I believe you want to simulate the situation where an instance of `Scrabble::Player` actually plays three words. What method should you call to set that up (and thus populate the `@plays` array)?

File: spec/lib/scrabble_spec.rb [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    end
+
+    [
+      ["highest", ["a", "an", "dog", "highest"]],
+      ["zipper", ["tan", "zipper", "pants", "hi", "cat"]]
+    ].each do |highest, array_of_words|
+      it "returns #{highest} from #{array_of_words}" do
+        expect(Scrabble::Scrabble.highest_score_from(array_of_words)).to eq(highest)
+      end
+    end
```
Comment (Elffers) [inline spec/lib/scrabble_spec.rb:51]: This is some clever metaprogramming, but I tend to find metaprogrammed tests harder to read. Since this is a unit test for a single method (i.e. `.highest_score_from`), I usually prefer writing one test that describes the functionality of that method. That is, instead of saying something like `it "returns X word from specific array Y" do`, you'd simply say, `it "returns the highest scoring word from an array of word" do`. Then, within that `it` block you can have two `expect` blocks if you really wanted to test more than one, though a single one should usually be enough. If there are unusual sets of word arrays, those should be tested separately as edge cases, with appropriate descriptions.

**Comment by sallyamoore** [comment]
Thank you so much for your review! So helpful!!! I'm very sorry I didn't get back to you sooner... it fell off the radar. Last week got away from me. 

I hope to incorporate your comments and make a new commit. I know it's been way too long to ask you to review my changes, so no need to do so! 

Thanks again!
